Add AWS SigV4 authentication support

Xtansia opened the pull request:
### Description
Implements authenticating requests via AWS SigV4 request signing.
 
I've taken a first attempt at this using the crate at https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-sigv4. I'm open to any feedback or improvements you might be able to offer on this. Main piece I'm unsure of would be the error handling, so if you can let me know what the desired manner would be, that'd be awesome.

### Issues Resolved
#36
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

rosscanning commented:
I originally requested this feature, and I can confirm that Xtansia's feature branch works for me.

dblock commented:
@rosscanning Could you please amend your last commit with `-s` for DCO?

dblock commented:
Let's document this, at least in README? And thanks!

rosscanning commented:
@dblock I haven't done any commits. I just requested the feature; @Xtansia did the actual work :)

dblock commented:
> @dblock I haven't done any commits. I just requested the feature; @Xtansia did the actual work :)

Yes, sorry I don't know what I was copy-pasting. Thanks @Xtansia !

Looks like we're seeing some test failures.

Xtansia commented:
@dblock I've fixed that commit and added some further documentation. Will take a look into the test failures. I don't believe I've changed anything that would cause them to fail.

mpetri commented on opensearch/src/http/aws_auth.rs:
```diff
@@ -0,0 +1,82 @@
+    credentials: &'a Credentials,
+    region: &'a Region,
+) -> Result<SigningParams<'a>, BuildError> {
+    let mut settings = SigningSettings::default();
```
could this be 

```
let settings = SigningSettings {
 percent_encoding_mode : PercentEncodingMode::Double;
 payload_checksum_kind : PayloadChecksumKind::NoHeader;
 signature_location : SignatureLocation::Headers;
 expires_in : None, 
 ..Default::default()
};

Xtansia replied:
`SigningSettings` is non_exhaustive, so unfortunately can not be constructed with a struct expression.

Xtansia replied:
Though on closer inspection the values I'm setting are the defaults, so will simplify it.

mpetri commented on opensearch/src/http/aws_auth.rs:
```diff
@@ -0,0 +1,82 @@
+        .settings(settings);
+
+    if let Some(session_token) = credentials.session_token() {
+        builder = builder.security_token(session_token);
```
this could be replaced with `builder.set_security_token(credentials.session_token())` ?

mpetri replied:
alternatively should this output an error message if this is `None`?

Xtansia replied:
Yes, will replace it 👍 The `None` state of session_token is fine, as permanent credentials won't have one, while temporary ones such as EC2 instance credentials etc will have one.

mpetri commented on opensearch/src/http/aws_auth.rs:
```diff
@@ -0,0 +1,82 @@
+
+    if let Some(new_headers) = instructions.take_headers() {
+        for (name, value) in new_headers.into_iter() {
+            request.headers_mut().insert(name.unwrap(), value);
```
should this either be `expect` or return an error for the whole function?

Xtansia replied:
I can change this to `expect`. In practice this is never `None`, it's just due to the API of `HeaderMap`. This is exactly what the aws-sigv4 crate does when operating on an `http::Request` https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-sigv4/src/http_request/sign.rs#L135-L139

mpetri replied:
yeah I understand it is never none. Maybe that is just my personal preference to never have `unwrap` in the code without a comment and then the comment might as well be in the `expect`

mpetri commented on opensearch/src/auth.rs:
```diff
@@ -45,6 +45,13 @@ pub enum Credentials {
+    /// 
+    /// This requires the `aws-auth` feature to be enabled.
+    #[cfg(feature = "aws-auth")]
+    Aws(aws_types::credentials::SharedCredentialsProvider, aws_types::region::Region)
```
should this be `AwsSigV4` in case there is another Aws specific auth method?

Xtansia replied:
Happy to rename this. Should I rename the feature as well? Or keep a "generic" `aws-auth` feature and just make the `Credentials` variant more specific?

mpetri replied:
Hmm not sure what is better. If I was using this I would expect the `aws-auth` feature to enable all aws specific auth methods so maybe just making the Credentials variant more specific is fine?

mpetri reviewed (changes requested):
Provided some general comments of some minor improvements I can see.

Xtansia commented:
@dblock @mpetri I believe I've addressed all the comments so far. Additionally it appears the integration test failure was due to `simple_logger` or `time` crate versions, so I've updated the version used by `yaml_test_runner`. This failure was unrelated to my changes as it failed the check on this commit 2 months ago as well for the same reason https://github.com/opensearch-project/opensearch-rs/actions/runs/1844072847

mpetri commented:
Looks okay to me!

dblock commented on opensearch/src/lib.rs:
```diff
@@ -58,6 +58,7 @@
 //! - **experimental-apis**: Enables experimental APIs. Experimental APIs are just that - an experiment. An experimental
```
Is this some generated file that should be excluded from the repo?

dblock replied:
Nm, ignore me. Rust noob ;)

dblock commented:
> Thank you @Xtansia for adding this feature!
> 
> Can we also add this usage in the [USER_GUIDE](https://github.com/opensearch-project/opensearch-rs/blob/main/USER_GUIDE.md)? It can be in another PR.

Good point. Let's do that as part of https://github.com/opensearch-project/opensearch-rs/issues/42.
